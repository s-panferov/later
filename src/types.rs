use chrono::Duration;

#[derive(Debug)]
pub enum Dimension {
	Second,
	Minute,
	Hour,
	Day,
	Week,
	Month,
	Quarter,
	Year,
}

#[derive(Debug)]
pub enum Period {
	Fixed(Duration),
	Month(i32),
	Quarter(i32),
	Year(i32),
}

#[derive(Debug)]
pub enum IntervalSpecifier {
	Period(Period),
}
