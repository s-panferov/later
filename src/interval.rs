use chrono::{DateTime, Duration, Utc};
use std::fmt::Debug;

#[derive(Debug, Clone)]
pub struct Interval {
	pub from: DateTime<Utc>,
	pub to: Option<DateTime<Utc>>,
}

impl Interval {
	pub fn from_now() -> Self {
		Interval {
			from: Utc::now(),
			to: None,
		}
	}

	pub fn from(from: DateTime<Utc>) -> Self {
		Interval { from, to: None }
	}
}

/// Something that resolves to the range of time
pub trait AsIntervals: Debug {
	/// Get the estimate of the interval
	fn duration_hint(&self) -> Duration;

	/// Get an interator to resolve intervals
	fn iter_within(
		&self,
		interval: Interval,
	) -> Box<dyn Iterator<Item = Interval>>;
}
