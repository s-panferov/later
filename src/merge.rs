use crate::interval::Interval;
use chrono::Duration;

pub use crate::merge_n::MergeN;

/// An iterator that merges consequest periods of time emitted by
/// another iterator.
pub struct Merge<T: Iterator<Item = Interval>> {
	prev: Option<Interval>,
	base: T,
	diff: Duration,
	completed: bool,
	max_merges: usize,
	merges: usize,
}

impl<T: Iterator<Item = Interval>> Merge<T> {
	pub fn new(base: T, diff: Duration, max_merges: usize) -> Self {
		Merge {
			base,
			prev: None,
			diff,
			completed: false,
			max_merges,
			merges: 0,
		}
	}
}

impl<T: Iterator<Item = Interval>> Iterator for Merge<T> {
	type Item = Interval;

	fn next(&mut self) -> Option<Interval> {
		if self.completed {
			return None;
		}

		self.merges = 0;

		loop {
			if self.merges > self.max_merges {
				self.completed = true;
				return None;
			}

			let value = self.base.next();
			if value.is_none() {
				self.completed = true;
				return std::mem::replace(&mut self.prev, None);
			}

			let value = value.unwrap();

			match &mut self.prev {
				None => {
					if value.to.is_none() {
						// all other intervals will be merged into this one anyway
						self.completed = true;
						return Some(value);
					} else {
						self.prev = Some(value);
					}
				}
				Some(prev) => {
					let to = prev.to.as_ref().unwrap();
					if value.from < *to || value.from - *to <= self.diff {
						// merge intervals
						prev.to = value.to;
						self.merges += 1;
						continue;
					} else {
						return std::mem::replace(&mut self.prev, Some(value));
					}
				}
			}
		}
	}
}

#[cfg(test)]
mod tests {
	use super::*;
	use crate::combine::Combine;
	use crate::interval::AsIntervals;
	use chrono::Weekday;
	use insta::assert_debug_snapshot_matches;

	#[test]
	fn merge_weekend() {
		let stream = Combine::new(vec![
			Weekday::Sat.iter_within(Interval::from(
				"2019-01-01T00:00:00Z".parse().unwrap(),
			)),
			Weekday::Sun.iter_within(Interval::from(
				"2019-01-01T00:00:00Z".parse().unwrap(),
			)),
		]);

		let stream = Merge::new(stream, Duration::seconds(1), 100);
		let what: Vec<Interval> = stream.take(10).collect();
		assert_debug_snapshot_matches!("merge_weekend", what);
	}
}
