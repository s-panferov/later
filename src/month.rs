use crate::interval::Interval;
use crate::utils::days_in_month;
use chrono::{Datelike, TimeZone, Timelike};

pub struct MonthIterator {
	interval: Interval,
	start_day: u32,
}

impl MonthIterator {
	pub fn new(interval: Interval) -> MonthIterator {
		let start_day = interval.from.day();
		MonthIterator {
			interval,
			start_day,
		}
	}
}

impl Iterator for MonthIterator {
	type Item = Interval;
	fn next(&mut self) -> Option<Interval> {
		let from = self.interval.from.clone();
		let month = from.month();

		let next_year = if month == 12 {
			from.year() + 1
		} else {
			from.year()
		};

		let next_month = if month == 12 { 1 } else { month + 1 };

		let days_in_next_month = days_in_month(next_year, next_month);
		let to = from;
		let next_day = if days_in_next_month < self.start_day {
			days_in_next_month
		} else {
			self.start_day
		};

		let to = to
			.offset()
			.ymd(next_year, next_month, next_day)
			.and_hms_nano(to.hour(), to.minute(), to.second(), to.nanosecond());

		if self.interval.to.is_some()
			&& to > *self.interval.to.as_ref().unwrap()
		{
			return None;
		}

		self.interval.from = to;

		Some(Interval { from, to: Some(to) })
	}
}
