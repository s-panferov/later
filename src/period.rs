use crate::interval::{AsIntervals, Interval};
use crate::types::Period;
use chrono::Duration;

use crate::merge::MergeN;
use crate::month::MonthIterator;

impl AsIntervals for Period {
	fn duration(&self) -> Duration {
		match self {
			Period::Fixed(d) => d.clone(),
			Period::Month(n) => Duration::weeks(4) * *n,
			Period::Quarter(n) => Duration::weeks(4) * 3 * *n,
			Period::Year(n) => Duration::weeks(4) * 12 * *n,
		}
	}

	fn iter_within(
		&self,
		interval: Interval,
	) -> Box<dyn Iterator<Item = Interval>> {
		match self {
			Period::Fixed(d) => d.iter_within(interval),
			Period::Month(m) => {
				Box::new(MergeN::new(MonthIterator::new(interval), *m as usize))
			}
			Period::Quarter(q) => Box::new(MergeN::new(
				MonthIterator::new(interval),
				(q * 3) as usize,
			)),
			Period::Year(y) => Box::new(MergeN::new(
				MonthIterator::new(interval),
				(y * 12) as usize,
			)),
		}
	}
}

#[cfg(test)]
mod tests {
	use super::*;
	use crate::interval::{AsIntervals, Interval};
	use insta::assert_debug_snapshot_matches;

	#[test]
	fn every_month() {
		let every_month = Period::Month(1);

		assert_debug_snapshot_matches!(
			"every_month",
			every_month
				.iter_within(Interval::from(
					"2019-10-31T12:00:00Z".parse().unwrap(),
				))
				.take(10)
				.collect::<Vec<_>>()
		);
	}
}
