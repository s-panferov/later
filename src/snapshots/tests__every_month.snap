---
created: "2019-08-31T23:26:51.551540Z"
creator: insta@0.10.1
source: src/period.rs
expression: "every_month.iter_within(Interval::from(\"2019-10-31T12:00:00Z\".parse().unwrap())).take(10).collect::<Vec<_>>()"
---
[
    Interval {
        from: 2019-10-31T12:00:00Z,
        to: Some(
            2019-11-30T12:00:00Z,
        ),
    },
    Interval {
        from: 2019-11-30T12:00:00Z,
        to: Some(
            2019-12-31T12:00:00Z,
        ),
    },
    Interval {
        from: 2019-12-31T12:00:00Z,
        to: Some(
            2020-01-31T12:00:00Z,
        ),
    },
    Interval {
        from: 2020-01-31T12:00:00Z,
        to: Some(
            2020-02-29T12:00:00Z,
        ),
    },
    Interval {
        from: 2020-02-29T12:00:00Z,
        to: Some(
            2020-03-31T12:00:00Z,
        ),
    },
    Interval {
        from: 2020-03-31T12:00:00Z,
        to: Some(
            2020-04-30T12:00:00Z,
        ),
    },
    Interval {
        from: 2020-04-30T12:00:00Z,
        to: Some(
            2020-05-31T12:00:00Z,
        ),
    },
    Interval {
        from: 2020-05-31T12:00:00Z,
        to: Some(
            2020-06-30T12:00:00Z,
        ),
    },
    Interval {
        from: 2020-06-30T12:00:00Z,
        to: Some(
            2020-07-31T12:00:00Z,
        ),
    },
    Interval {
        from: 2020-07-31T12:00:00Z,
        to: Some(
            2020-08-31T12:00:00Z,
        ),
    },
]
